cirq-core/cirq/transformers/transformer_api.py

```diff
@@ -300,14 +264,45 @@ def transformer(cls_or_func: Any) -> Any:
     >>>        ...
 
     Args:
-        cls_or_func: The callable class or method to be decorated.
+        cls_or_func: The callable class or function to be decorated.
 
     Returns:
-        Decorated class / method which includes additional logging boilerplate. The decorated
-        callable always receives a copy of the input circuit so that the input is never mutated.
+        Decorated class / function which includes additional logging boilerplate.
     """
     if isinstance(cls_or_func, type):
-        return _transformer_class(cls_or_func)
+        cls = cls_or_func
+        method = cls.__call__
+
+        @functools.wraps(method)
+        def method_with_logging(self, circuit, context):
```

Comment:
Why don't we need any type annotations here (and below in the func case) ?

Reply:
Basically, by omitting type annotations here we are telling mypy not to typecheck the body of this function. We still get checking of the `transformer` decorator itself, just not all the internals. IMO this is worth it because the implementation is simple and the gymnastics to get mypy to check everything seems to obscure what is actually going on. But if you think this goes too far I can add some annotations back in.

Reply:
Unless and until it doesn't affect the type annotations for the decorated method / class and types for the transformer call sites work as expected, I'm fine with removing the annotations in the internal implementation. 

To make sure I understand correctly -- We are telling mypy that the decorator will preserve the input type (using `_TRANSFORMER_T` and `_TRANSFORMER_CLS_T` in this case) and that's enough to make sure type annotations for both the decorated methods and the call sites work correctly. Beyond this, we don't really care about the types in the internal implementation.

Reply:
Added in basic annotations (`cirq.AbstractCircuit` -> `cirq.AbstractCircuit`) and everything still type checks (I assume this is because of the `Any` -> `Any` annotations on the final transformer implementation. I checked that the revealed type of decorated functions still preserves the specific output circuit type, so basically what's happening is that the overloads specify the _true_ type contract of the `transformer` decorator, even though mypy can't check this fully.